src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.DecodeRecipients.cs

```diff
@@ -0,0 +1,306 @@
+
+            if (encodedKeyAgreement.PeekTag() == contextSpecific1)
+            {
+                // Skip extra keying material; this one isn't exposed in the API and Framwork does
+                // nothing with it.
+                encodedKeyAgreement.SkipValue();
+            }
+
+            AlgorithmIdentifier keyEncryptionAlgo = encodedKeyAgreement.ReadAlgoIdentifier();
+
+            // As there might be more than one element in RecipientEncryptedKeys, we have to create one KeyAgreeRecipientInfo for each
+            // one of the elements. 
+            List<SubjectIdentifier> recipientIdSet = new List<SubjectIdentifier>();
+            List<CryptographicAttributeObject> otherKeyMaterialSet = new List<CryptographicAttributeObject>();
+            List<DateTime> dateSet = new List<DateTime>();
+            List<byte[]> encryptionKeySet = new List<byte[]>();
+
+            ReadRecipientEncryptedKeys(encodedKeyAgreement, recipientIdSet, otherKeyMaterialSet, dateSet, encryptionKeySet);
+
+            List<KeyAgreeRecipientInfo> keyAgreeRecipients = new List<KeyAgreeRecipientInfo>(recipientIdSet.Count);
+            for (int i = 0; i < recipientIdSet.Count; i++)
+            {
+                KeyAgreeRecipientInfoPalOpenSsl kari =
+                    new KeyAgreeRecipientInfoPalOpenSsl(
+                        version,
+                        recipientIdSet[i],
+                        originatorIdentifierOrKey,
+                        otherKeyMaterialSet[i],
+                        keyEncryptionAlgo,
+                        encryptionKeySet[i],
+                        dateSet[i]);
+                keyAgreeRecipients.Add(new KeyAgreeRecipientInfo(kari));
+            }
+            return keyAgreeRecipients;
+        }
+
+        private static PublicKeyInfo ReadOriginatorPublicKey(DerSequenceReader encodedKeyAgreement)
+        {
+            // OriginatorPublicKey::= SEQUENCE {
+            //      algorithm AlgorithmIdentifier,
+            //      publicKey BIT STRING }
+            DerSequenceReader publicKeyReader = encodedKeyAgreement.ReadSequence();
+
+            DerSequenceReader algoIdentifier = publicKeyReader.ReadSequence();
+
+            Oid algoOid = algoIdentifier.ReadOid();
+            byte[] publicKey = publicKeyReader.ReadBitString();
+ 
+            // This section reads the key and determines the bit length of the key
+            byte[] rawKey = DerSequenceReader.CreateForPayload(publicKey).ReadIntegerBytes();
+            int trailingZeros = 0;
+            while (rawKey[trailingZeros++] == 0);
+            int keyLength = 8 * (rawKey.Length - trailingZeros);
+
+            AlgorithmIdentifier publicKeyAlgorithm = new AlgorithmIdentifier(algoOid, keyLength);
+
+            return new PublicKeyInfo(publicKeyAlgorithm, publicKey);
+        }
+
+        private static KeyTransRecipientInfo ReadKeyTrans(DerSequenceReader recipientSet)
+        {
+            // As of RFC 2630 this is a KeyTransRecipientInfo 
```

Comment:
I'm not sure what the "As of...this is" means.  Maybe "As of RFC 2630 a KeyTransRecipientInfo is"?  The comment as-written makes me think it got renamed, or something.